Documentation changes

quantumgc opened the pull request:
This PR changes the inline citations to use DocumenterCitations.jl.
It resolves #37.

This branch currently tracks [DisneyHogg:readying_paper_for_submission](https://github.com/DisneyHogg/Skyrmions3D.jl/tree/readying_paper_for_submission) since that contains the bibliography information.
Accordingly, this PR must not be merged into main until the JOSS submission branch is.

DisneyHogg commented on docs/src/making_skyrmions.md:
```diff
@@ -50,7 +50,7 @@ The `make_rational_map!` function also accepts custom profile functions.
 ## ADHM data
 
-[ADHM skyrmions](https://arxiv.org/pdf/2110.15190) are skyrmions generated from ADHM data. The data consists of symmetric quaternionic matrices which satisfy a constraint. Most highly symmetric skyrmions can be represented by ADHM data. This package implements the very efficient [parallel transport](https://arxiv.org/abs/2204.04032) algorithm from Derek Harland. We can make the baryon number 2 to ...
+ADHM skyrmions [corkADHMSkyrmions2022](@cite) are skyrmions generated from ADHM data. The data consists of symmetric quaternionic matrices which satisfy a constraint. Most highly symmetric skyrmions can be represented by ADHM data. This package implements the very efficient parallel transport algorithm developed by [Harland2023](@citet). We can make the baryon number 2 toroidal skyrmion as follow ...
```
This looks like a simple typo, `citet` where it should be `cite`

quantumgc replied:
This is intentional, `citet` prints the author as opposed to just inserting a (numeric) reference.

quantumgc committed:
skip doctests

Commit:
set doctests to warn

Commit:
fix typos

Commit:
add doctest setup blocks

Add @meta blocks around each subsection which import modules needed by
the various doctests.

Commit:
split example block and add expl. text + doctests

Split up the large example block which details how to initialise vacuum
skyrmions. This improves legibility slightly and allows for some
explanatory text to be added in between.

Also convert the julia blocks to doctest blocks.

Commit:
minor edits

- edit last sentence of subsection
- add crossrefs to `Skyrmion` constructor
- edit intro para of rat map sec

Commit:
convert rat map code blocks to doctests

Commit:
edits to rational maps section

- insert explanation of how skyrmions are made
- edit explanatory text about profile function finding and baryon number

Commit:
begin section on custom profile funcs

Commit:
add doctests to ADHM data section

- add quaternions dependency to docs
- convert code block to doctest
- add energy computation

Commit:
add filters to doctests

Filter float results so that only the first four digits after the
decimal point are compared.
Prevents doctests failing due to differening FPA implementations on
testing machines.

quantumgc commented:
I've made some changes to this PR since it seems simpler to collect the documentation changes in one place rather than splitting them up. I've also made a fix (see e33e843) so this can be merged independently of the paper branch (the fix can be undone relatively easily later on).

The main planned changes for this PR are:
- change the citations to use DocumenterCitations.jl (resolving #37).
- add doctests to the code examples (resolving #38).
- edit the documentation to bring it up to spec with the package as it is now.

quantumgc commented:
The todo list for this PR is currently:
+ [ ] Creating Skyrmions
  + [x] Add code examples for custom profile functions
  + [x] Verify energies
  + [ ] Doctests for saving and loading skyrmions
+ [ ] Transform and Combine
+ [ ] Compute properties
+ [ ] Flow
+ [ ] Visualise

Commit:
hotfix: add id to properties page

This allows the tutorials page to be crossrefed, but I don't think it's
the best way of doing this. Needs to be revisited

Commit:
typos

- fix charge in name of b_3_tet skyrmion
- fix charge in name of b_2_tor_skyrmion

Commit:
minor edits

Commit:
add example of custom profile function

Commit:
add explicit using statement in examples

DisneyHogg commented on docs/src/flow.md:
```diff
@@ -52,13 +54,16 @@ Note that, depending on your boundary conditions and lattice size, your specifie
 ## Arrested Newton Flow
 
-A more efficient algorith is the Newton Flow method. This models the second order differential equation
+A more efficient algorith is the Newton Flow method.
```
typo, should be "algorithm"

DisneyHogg commented on docs/src/index.md:
```diff
@@ -8,26 +8,38 @@ A [Julia](http://julialang.org) package for creating, manipulating and probing S
 ### Authors
 
-I am Chris Halcrow, a research software engineer at the University of Edinburgh. I would love there to be more authors of this package. Please join in.
+I am Chris Halcrow, a research software engineer at the University of Edinburgh.
```
As we're making documentation changes we could add something link 
"
I've been supported in development by Linden Disney-Hogg and Cas G. Chaudhuri.
"

quantumgc replied:
Do we want to hyperlink the names to our respective websites or should we leave them unadorned?

chrishalcrow replied:
Oh, let's make this a bit more equal. Maybe:

The current core contributors are:
- Chris Halcrow, a research software engineer at the University of Edinburgh.
- etc

quantumgc commented:
Rebased onto main to incorporate changes from #40

chrishalcrow commented:
Not really been following this (sorry). Looks great!!

chrishalcrow commented:
I did some aggresive editing... I can't seem to commit directly onto this PR, so here's the md file after lotsa changes. Happy to debate anything:

[paper.md](https://github.com/user-attachments/files/23423514/paper.md)

chrishalcrow commented:
This is looking good. Do you need to work on it more @quantumgc ?

Commit:
update documentation dependencies

DisneyHogg commented on docs/src/visualisation.md:
````diff
@@ -21,15 +23,17 @@ We can plot the field with component `3` and isosurface with `iso_value=0.3` as
+Hopefully whatever happens, you can see the following plot:
 
 ```@raw html
 <img src="../../../src/images/visualisation/B3_field.png>
````
another inverted commas missing here?

DisneyHogg commented on docs/make.jl:
```diff
@@ -1,4 +1,9 @@
+using Documenter, Skyrmions3D, DocumenterCitations
+
+bib = CitationBibliography(
+    joinpath(@__DIR__, "src/ref.bib");
```
are you able to change this now to `"paper/paper.bib"` because the paper PR has been merged?

DisneyHogg commented:
It seems like the job of adding doctests to the documentation is the major blocker to this PR at the moment, and that in turn is blocking the JOSS submission? Should we thus split off the job of adding the doctests to a separate PR, merge this one (modulo fixing the existing comments) and then we can make the submission.

quantumgc commented:
Hi both,

Sorry for not touching this project for a while.

I've just submitted #45 for review which contains the changes needed to add DocumenterCitations.jl, that can be merged without issue. I think it makes sense to either close or revert this PR to WIP status, since adding the doctests is turning out to be a non-trivial task. I'll add a summary of where I'm at right now and the problems I'm facing below.

I probably won't have time to look at this again until after I've submitted the thesis, so do feel free to work on it in my absence.

## Status quo

Doctests have been written for all sections of `making_skyrmions.md` besides the saving and loading skyrmions portion. They have also been written for the translation portion of `transform.md`, though these might be subject to change later. `flow.md`, `properties.md`, and `visualisation.md` have not been touched, yet I think it should be a fairly simple procedure to add doctests to the properties page.

## Issues

### Verifying correctness of `Vector{Float}` and negative zero outputs

Documenter.jl has a feature where you can [filter](https://documenter.juliadocs.org/stable/man/doctests/#Filtering-Doctests) the output of a documentation block with a regex. This can be used to e.g. filter out parts of a `Float` so we only check that an answer is correct to machine epsilon (this is used in the `making_skyrmions.md` section).

This works fine for single `Float`s, but I'm having difficulty applying this feature to `Vector{Float}` types. [`round`](https://docs.julialang.org/en/v1/base/math/#Base.round)ing any such output to a desired level of accuracy works, yet it seems somewhat inelegant as the filtering should be strong enough to handle this.

I've also run into similar problems when an output is negative zero, the regexs I've written can't seem to detect and change those.

### Demonstrating what functions do

In `transform.md`, the usage patterns of functions are given, but their effect is not demonstrated in the documentation. This presents an issue for documentation testing as Documenter.jl does not just verify that a code block runs, it must also verify that a code block returns a certain output. I've edited the text and the documentation examples to demonstrate what some of the functions do (e.g. the `translate_sk` function, I've also done something similar locally for the `rotate_sk` function), but I'm having some difficulties writing example code for the other functions.

### Presentation

As it stands, the code blocks are a mix of styles (either REPL style or code block style). I've done my best to keep to whatever style was originally present in the documentation, but the lack of consistency can be a bit distracting. Is there a preferred style for the entire documentation, or a rationale as to which to use in a particular situation?
